### PR TITLE
HPCC-30028 LDAP connection pool improvement when host down

### PR DESCRIPTION
**Pull request by RussWhitehead**
Make connection requests fail over to next AD faster when current one goes down. Immediately blacklist current AD when LDAP_SERVER_DOWN returned. Check for current AD being blacklisted before attempting retry

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by RussWhitehead** [comment]
@kenrowland Please review

**Comment by RussWhitehead** [comment]
@kenrowland  FYI, I tested with 2  instances of 389DirectoryServer, and randomly brought them up and down whilst performing ECLWatch activities that requires LDAP.  Note this fix will significantly speed up primary and secondary connection failovers, since we immediately blacklist an AD when LDAP returns the SERVER DOWN error (prior code kept retrying)

File: system/security/LdapSecurity/ldapconnection.cpp [hunk 1]
```diff
@@ -983,44 +983,33 @@ class CLdapConnection : implements ILdapConnection, public CInterface
...
-            }
-
-            if(rc == LDAP_SERVER_DOWN)
-            {
-                StringBuffer dc;
-                LdapUtils::getDcName(m_ldapconfig->getDomain(), dc);
-                if(dc.length() > 0)
+                m_ldapconfig->getLdapHost(ldapHost);//@@get currently selected host, which may change
+                if(retries)
+                    DBGLOG("LDAP Server temporarily unreachable, retrying on %s...", ldapHost.str());
```
Comment (kenrowland) [inline system/security/LdapSecurity/ldapconnection.cpp:993]: Why not add the retry number?

File: system/security/LdapSecurity/ldapconnection.cpp [hunk 2]
```diff
@@ -983,44 +983,33 @@ class CLdapConnection : implements ILdapConnection, public CInterface
...
+                if(rc == LDAP_SUCCESS)
+                    return true;
+                else if(rc == LDAP_SERVER_DOWN)
                 {
-                    WARNLOG("Using automatically obtained LDAP Server %s", dc.str());
-                    rc = connect(dc.str(), proto);
+                    if (m_ldapconfig->getHostCount())
+                    {
+                        m_ldapconfig->rejectHost(ldapHost);
+                        retries = 0;//make sure this host gets adequate number of attempts
```
Comment (kenrowland) [inline system/security/LdapSecurity/ldapconnection.cpp:1002]: Is it possible that resetting to 0 will cause the retry loop to never exit? 

update: based on not being able to reject a host when there is only one available, it looks like this could cause the code to execute in a loop until the single host is available.

File: system/security/LdapSecurity/ldapconnection.cpp [hunk 3]
```diff
@@ -983,44 +983,33 @@ class CLdapConnection : implements ILdapConnection, public CInterface
...
+                if(retries)
+                    DBGLOG("LDAP Server temporarily unreachable, retrying on %s...", ldapHost.str());
+                rc = connect(ldapHost.str(), proto);
+                if(rc == LDAP_SUCCESS)
+                    return true;
+                else if(rc == LDAP_SERVER_DOWN)
                 {
-                    WARNLOG("Using automatically obtained LDAP Server %s", dc.str());
-                    rc = connect(dc.str(), proto);
+                    if (m_ldapconfig->getHostCount())
```
Comment (kenrowland) [inline system/security/LdapSecurity/ldapconnection.cpp:999]: I am guessing you check this number because it can change instead of assuming it's still valid from the outermost loop. However, looking at how the number is managed wrt to rejecting a host, it won't go to 0. Since it's always positive (is the check needed), is there a case where you should actually sleep (when only one host and you can't temporarily reject it)?

File: system/security/LdapSecurity/ldapconnection.cpp (hunk 1 above)
Comment (RussWhitehead) [reply system/security/LdapSecurity/ldapconnection.cpp:993]: Done

File: system/security/LdapSecurity/ldapconnection.cpp (hunk 3 above)
Comment (RussWhitehead) [reply system/security/LdapSecurity/ldapconnection.cpp:999]: Reworked logic makes this go away

**Review by kenrowland — APPROVED** [review]

**Comment by RussWhitehead** [comment]
@ghalliday Please merge this vault improvement into 8.12.x

File: system/security/LdapSecurity/ldapconnection.cpp [hunk 4]
```diff
@@ -983,44 +983,25 @@ class CLdapConnection : implements ILdapConnection, public CInterface
...
-                if(!LdapServerDown(rc) || retries > LDAPSEC_MAX_RETRIES)
+                m_ldapconfig->getLdapHost(ldapHost);//get currently selected host, which may change
+                if(retries)
+                    DBGLOG("LDAP Server unreachable, retry attempt %d of %d on %s...", retries, LDAPSEC_MAX_RETRIES, ldapHost.str());
+                rc = connect(ldapHost.str(), proto);
+                if(rc == LDAP_SUCCESS)
+                    return true;
+                else if(rc == LDAP_SERVER_DOWN)
                     break;
                 sleep(LDAPSEC_RETRY_WAIT);
```
Comment (ghalliday) [inline system/security/LdapSecurity/ldapconnection.cpp:999]: There is no benefit in sleeping if the number of retries has expired.  Previously it would have exited the loop before sleeping.

**Review by ghalliday — CHANGES_REQUESTED** [review]
One minor comment related to retires.
Also there are 2 other sets of code (authenticate, and CLdapConfig())  that have similar loops.  Do they need similar changes?

As a separate PR I suspect the number of retries and the sleep should be configurable.

File: system/security/LdapSecurity/ldapconnection.cpp [hunk 5]
```diff
@@ -465,16 +464,17 @@ class CLdapConfig : implements ILdapConfig, public CInterface
 
             for(int retries = 0; retries <= LDAPSEC_MAX_RETRIES; retries++)
             {
+                getLdapHost(hostbuf);
```
Comment (ghalliday) [inline system/security/LdapSecurity/ldapconnection.cpp:467]: Why move this line?  It now will read it each time round the retry loop?

File: system/security/LdapSecurity/ldapconnection.cpp [hunk 6]
```diff
@@ -465,16 +464,17 @@ class CLdapConfig : implements ILdapConfig, public CInterface
...
                 rc = LdapUtils::getServerInfo(hostbuf.str(), sysUserDN.str(), m_sysuser_password.str(), m_protocol, port, m_cipherSuite, dcbuf, m_serverType, ldapDomain, m_timeout);
-                if(!LdapServerDown(rc) || retries >= LDAPSEC_MAX_RETRIES)
-                    break;
-                sleep(LDAPSEC_RETRY_WAIT);
-                if(retries < LDAPSEC_MAX_RETRIES)
+                if(rc == LDAP_TIMEOUT && retries <= LDAPSEC_MAX_RETRIES)
```
Comment (ghalliday) [inline system/security/LdapSecurity/ldapconnection.cpp:469]: This seems to deliberately be making the code as hard to follow as possible.  Why not

```
                if(rc != LDAP_TIMEOUT || retries >= LDAPSEC_MAX_RETRIES)
                    break;
                  sleep(LDAPSEC_RETRY_WAIT);
                  DBGLOG("LDAP AD Server %s temporarily unreachable for user %s on port %d, retrying...", hostbuf.str(), sysUserDN.str(), port);
```
Your code is also wrong - it will sleep even if it has matched the maximum retries.

File: system/security/LdapSecurity/ldapconnection.cpp [hunk 7]
```diff
@@ -465,16 +464,17 @@ class CLdapConfig : implements ILdapConfig, public CInterface
...
                 {
+                    sleep(LDAPSEC_RETRY_WAIT);
                     DBGLOG("LDAP AD Server %s temporarily unreachable for user %s on port %d, retrying...", hostbuf.str(), sysUserDN.str(), port);
+                    continue;
                 }
+                break;
             }
-            if (rc != LDAP_SUCCESS)
+            if(LdapServerDown(rc))//LDAP_SERVER_DOWN || LDAP_UNAVAILABLE || LDAP_TIMEOUT
```
Comment (ghalliday) [inline system/security/LdapSecurity/ldapconnection.cpp:477]: The comments don't need to be there

File: system/security/LdapSecurity/ldapconnection.cpp [hunk 8]
```diff
@@ -986,41 +986,26 @@ class CLdapConnection : implements ILdapConnection, public CInterface
...
-                if(retries < LDAPSEC_MAX_RETRIES)
-                    DBGLOG("Server temporarily unreachable, retrying ...");
-            }
-
-            if(rc == LDAP_SERVER_DOWN)
-            {
-                StringBuffer dc;
-                LdapUtils::getDcName(m_ldapconfig->getDomain(), dc);
-                if(dc.length() > 0)
+                if(rc == LDAP_SUCCESS)
```
Comment (ghalliday) [inline system/security/LdapSecurity/ldapconnection.cpp:993]: Changed in a completely different way to the code above?  Why??

File: system/security/LdapSecurity/ldapconnection.cpp [hunk 9]
```diff
@@ -1968,16 +1952,24 @@ class CLdapClient : implements ILdapClient, public CInterface
...
                     break;
-                sleep(LDAPSEC_RETRY_WAIT);
-                if(retries < LDAPSEC_MAX_RETRIES)
-                    DBGLOG("Server temporarily unreachable, retrying ...");
-                // Retrying next ldap sever, might be the same server
-                m_ldapconfig->getLdapHost(hostbuf);
+                else if(rc==LDAP_SERVER_DOWN || rc==LDAP_UNAVAILABLE)
+                {
+                    m_ldapconfig->rejectHost(hostbuf);
```
Comment (ghalliday) [inline system/security/LdapSecurity/ldapconnection.cpp:1960]: Changed in a third different way?  I see that it isn't looping through the hosts in the same way, but I would expect the code to be changed to make it more similar.

**Review by ghalliday — CHANGES_REQUESTED** [review]

File: system/security/LdapSecurity/ldapconnection.cpp (hunk 5 above)
Comment (RussWhitehead) [reply system/security/LdapSecurity/ldapconnection.cpp:467]: That's the point, and the problem this Jira is fixing. A different thread may blacklist the currently selected host, in which case we always need to get the latest

File: system/security/LdapSecurity/ldapconnection.cpp (hunk 7 above)
Comment (RussWhitehead) [reply system/security/LdapSecurity/ldapconnection.cpp]: Removed

File: system/security/LdapSecurity/ldapconnection.cpp (hunk 8 above)
Comment (RussWhitehead) [reply system/security/LdapSecurity/ldapconnection.cpp:990]: The above code is executed at startup, and tries to connect to the first LDAP AD even while in the retry loop. This is the same as the legacy implementation. It will blacklist it after retrying. This code (connection pool) needs to be sensitive to an AD going down at any point, even in the retry loop.  This is where Lisa will see the improvement

File: system/security/LdapSecurity/ldapconnection.cpp (hunk 9 above)
Comment (RussWhitehead) [reply system/security/LdapSecurity/ldapconnection.cpp:1955]: Removed the unnecessary check for LDAP_SUCCESS, but note that the rejectHost call makes sense inside of the retry loop, since there isn't an outer loop going through numHosts

**Review by RussWhitehead — COMMENTED** [review]
@ghalliday Please rereview

File: system/security/LdapSecurity/ldapconnection.cpp (hunk 5 above)
Comment (ghalliday) [reply system/security/LdapSecurity/ldapconnection.cpp:467]: That needs a comment in the code because it is not clear from the code - especially given the way the outer loop is structured.

File: system/security/LdapSecurity/ldapconnection.cpp (hunk 8 above)
Comment (ghalliday) [reply system/security/LdapSecurity/ldapconnection.cpp:990]: Yes, but it is clearer if it is coded as:

```
               if (rc == LDAP_SUCCESS)
                    return true;
                if(rc != LDAP_TIMEOUT || retries >= LDAPSEC_MAX_RETRIES)
                    break;
                sleep(LDAPSEC_RETRY_WAIT);
                DBGLOG("LDAP AD Server %s temporarily unreachable for user %s on port %d, retrying...", hostbuf.str(), sysUserDN.str(), port);


```  
so the structure matches the structure of the 1st loop.

File: system/security/LdapSecurity/ldapconnection.cpp (hunk 9 above)
Comment (ghalliday) [reply system/security/LdapSecurity/ldapconnection.cpp:1955]: Why isn't there an outer loop?  I think it was missed as part of the changes in HPCC-16179.

It means that only 3 servers will be checked, but that shouldn't cause problems in practice.

File: system/security/LdapSecurity/ldapconnection.cpp [hunk 10]
```diff
@@ -986,41 +983,26 @@ class CLdapConnection : implements ILdapConnection, public CInterface
...
-                    break;
-                sleep(LDAPSEC_RETRY_WAIT);
-                if(retries < LDAPSEC_MAX_RETRIES)
-                    DBGLOG("Server temporarily unreachable, retrying ...");
-            }
-
-            if(rc == LDAP_SERVER_DOWN)
-            {
-                StringBuffer dc;
-                LdapUtils::getDcName(m_ldapconfig->getDomain(), dc);
```
Comment (ghalliday) [inline system/security/LdapSecurity/ldapconnection.cpp:1004]: Was this code was removed on purpose?  There is still similar code in authenticate().  Is that correct?

**Review by ghalliday — COMMENTED** [review]
I think this code is ok (although one question about deleted code).  There are a few comments on keeping similar code consistent/minimizing the changes.

A couple of other things that might be worth looking at in a separate PR
- Should the timeouts vary on the operation being performed?  E.g. I imagine getServerInfo() should return quickly.
- Some of the tracing looks verbose (e.g. two lines for each authenticate().  I don't know how often it occurs though.

File: system/security/LdapSecurity/ldapconnection.cpp [hunk 11]
```diff
@@ -1968,16 +1944,22 @@ class CLdapClient : implements ILdapClient, public CInterface
...
                 }
                 DBGLOG("finished LdapBind for user %s, rc=%d", username, rc);
-                if(!LdapServerDown(rc) || retries > LDAPSEC_MAX_RETRIES)
+
+                if(rc==LDAP_SERVER_DOWN || rc==LDAP_UNAVAILABLE)
+                {
+                    m_ldapconfig->rejectHost(hostbuf);
+                    continue;//try again with next configured LDAP host
+                }
+                else if(rc==LDAP_TIMEOUT && retries < LDAPSEC_MAX_RETRIES)
```
Comment (kenrowland) [inline system/security/LdapSecurity/ldapconnection.cpp:1953]: Same comment as above. Curious why the test on retries here and not let the loop exit test handle it.

File: system/security/LdapSecurity/ldapconnection.cpp [hunk 12]
```diff
@@ -465,16 +464,14 @@ class CLdapConfig : implements ILdapConfig, public CInterface
 
             for(int retries = 0; retries <= LDAPSEC_MAX_RETRIES; retries++)
             {
+                getLdapHost(hostbuf);//get next available AD, as it may have changed
                 rc = LdapUtils::getServerInfo(hostbuf.str(), sysUserDN.str(), m_sysuser_password.str(), m_protocol, port, m_cipherSuite, dcbuf, m_serverType, ldapDomain, m_timeout);
-                if(!LdapServerDown(rc) || retries >= LDAPSEC_MAX_RETRIES)
+                if(rc != LDAP_TIMEOUT || retries >= LDAPSEC_MAX_RETRIES)
```
[the remark this answers is not in the document]
Comment (ghalliday) [reply system/security/LdapSecurity/ldapconnection.cpp:469]: Because it prevents the sleep it is no longer going to retry that connection.

File: system/security/LdapSecurity/ldapconnection.cpp (hunk 11 above)
Comment (ghalliday) [reply system/security/LdapSecurity/ldapconnection.cpp:1953]: Same response as above.

**Review by ghalliday — APPROVED** [review]
Those changes look good.  The third loop could do with adding a host loop for consistency, but that can/should be another PR.
